### PR TITLE
Introduce package conditionals to Ansible remediations

### PR DESCRIPTION
**Pull request by yuumasato**
#### Description:

- Trickle down the CPE Platform to the Ansible remediations as `when` conditionals.
  - When a rule has a `platform` that is not `machine`, a `when` condition is added.
  - `fact_package` tasks are added to remediations that have a when condition with `ansible_facts.packages`

#### Rationale:
When a rule has a Package CPE platform, like `platform: grub2` for example.
It is meant to be evaluated and remediated when the  systems has grub2 package installed.
Evaluation and remediation done with OpenSCAP respects these CPEs, as the scanner understands the `CPE` and the `XCCDF:Platform` field.
But when the Playbooks are generated, the Package applicability of the remediation is lost.

### CODE REVIEW

**Comment by mildas** [comment]
**Changes identified:**
Others:
 Python abstract syntax tree change found in ssg/build_remediations.py.
 Python abstract syntax tree change found in ssg/build_yaml.py.

**Recommended tests to execute:**
 (cd build && cmake ../ && ctest -j4)

File: ssg/build_remediations.py [hunk 1]
```diff
@@ -343,10 +342,56 @@ def _get_rule_reference(self, ref_class):
         else:
             return []
 
+    def inject_package_facts_task(self, parsed_snippet):
+        """ Injects a package_facts task only if
+            the snippet has a task with a when clause with ansible_facts.packages,
+            and the snippet doesn't already have an package_facts task
```
Comment (jan-cerny) [inline ssg/build_remediations.py:348]:
```suggestion
            and the snippet doesn't already have a package_facts task
```

**Review by jan-cerny — COMMENTED** [review]
This is great and works fine for me.

I don't like the fact that it causes the gathering of packages repeatedly, for example, in RHEL 8  OSPP Playbooks there are 37 tasks "Gather the package facts". But I don't know if we're able to do something with that at this moment. I guess we would have to change the way the profile playbooks are generated from the datastreams. That would be a big separate effort.

File: ssg/build_remediations.py [hunk 2]
```diff
@@ -343,10 +342,56 @@ def _get_rule_reference(self, ref_class):
...
+        """
+        has_package_facts_task = False
+        has_ansible_facts_packages_clause = False
+
+        for p_task in parsed_snippet:
+            # We are only interested in the OrderedDicts, which represent Ansible tasks
+            if not isinstance(p_task, dict):
```
[the remark this answers is not in the document]
Comment (yuumasato) [reply ssg/build_remediations.py:355]: I would keep as `dict`.
An `OrderedDict` is also an instance of `dict`, but a `dict` is not an instance of an `OrderedDict`.

**Comment by mildas** [comment]
**Changes identified:**
Others:
 Python abstract syntax tree change found in ssg/build_remediations.py.
 Python abstract syntax tree change found in ssg/build_yaml.py.

**Recommended tests to execute:**
 (cd build && cmake ../ && ctest -j4)

**Comment by yuumasato** [comment]
@jan-cerny Thank you for the review, Honzo!

> I don't like the fact that it causes the gathering of packages repeatedly, for example, in RHEL 8 OSPP Playbooks there are 37 tasks "Gather the package facts". But I don't know if we're able to do something with that at this moment. I guess we would have to change the way the profile playbooks are generated from the datastreams. That would be a big separate effort.

I agree with you here, the aspect of repeated tasks needs to be addressed at some moment.
This can be tackled on the build system of the content, but I think this also needs to be addressed on the scanner side, so that it works for `oscap xccdf generate fix` too.

**Comment by jan-cerny** [comment]
@yuumasato Great!

Pending CI.

**Pull request merged**